# GEOPHYS 257 (Winter 2023)

## Data Types and Basic Machine Errors

In this lab we will be covering Numeric Data Types for both Python and Numpy, as well as Machine errors. Although the problems in this lab are not from *Python Numerical Methods*, please read [Chapter-9](https://pythonnumericalmethods.berkeley.edu/notebooks/chapter09.00-Representation-of-Numbers.html). Pay particular attention to the *Round-off Errors* secion, but the first two sections are mostly FYI; however, these sections provide an understanding of how most computational systems represent fractional numbers via binary numbers.

[//]: <> (Notebook Author: Thomas Cullison, Stanford University, Jan. 2023)

## External Resources
If you have any question regarding some specific Python functionality you can consult the official [Python documenation](http://docs.python.org/3/).

* [Python Numeric Types](https://docs.python.org/3/library/stdtypes.html#numeric-types-int-float-complex)
* [Numpy Data Types](https://numpy.org/doc/stable/user/basics.types.html)

## Python Built-in Numeric Types (not the same as Numpy types)

In Python it is not necessary to declare variables. Once a value is assigned to a variable the variable will have the type of that value. That behaviour is known as **dynamic typing**. A variable can also change its type at any point of the program execution. That has its advantages but there are also some pitfalls.

1. Int
1. Float
1. Complex (really the 3rd type is *imaginary* and the complex number is like a tuple)
1. Type Casting

## Exercise 0

#### First read [Chapter 9](https://pythonnumericalmethods.berkeley.edu/notebooks/chapter09.00-Representation-of-Numbers.html) in *Python Numerical Methods*.

Then begin this lab by first reading the code below; then running it, and, finally, by examing the results. Continue to the next exercise afterwards.

In [1]:
# 1
x=2
print('x =',x)
print('x type:',type(x))
print()

# 2 manual casting int to float
cx=float(x)
print('cx =',cx)
print('cx type:',type(cx))
print()

# 3
y=1.2
print('y =',y)
print('y type:',type(y))
print()

# 4 manual casting float to int
cy=int(y)
print('cy =',cy) # !!! notice the rounding !!!
print('cy type:',type(cy))
print()

# 3
z=1 + 2j
print('z =',z)
print('z type:',type(z))
print()

# 4 dynamic casting of y to complex
print('z*y =',z*y)
print('z*y type:',type(z*y))
print()

x = 2
x type: <class 'int'>

cx = 2.0
cx type: <class 'float'>

y = 1.2
y type: <class 'float'>

cy = 1
cy type: <class 'int'>

z = (1+2j)
z type: <class 'complex'>

z*y = (1.2+2.4j)
z*y type: <class 'complex'>



## Exercise 1: Dynamic Type Casting

You might have noticed that $z \cdot y$ above returned a complex number. This is related to something called *type casting*. The value stored in $y$ was actually casted (think transformed) into a complex number and then the multiplication was computed. Thus, the two lines below will yield equivalent answers.

```python 
z*y
z*(y + 0.j)
```

Note, the dot in '0.j' is needed to tell Python that I'm adding a floating point imaginary number. This dot is only neccessary when using '0' as the imaginary part, or when you wish to explicitly define a values as being a float type.

To confirm that the above two lines are equivelant, try the below line for yourself. 

```python 
z*y == z*(y + 0.j)
```

It will return a value of 'True'. The '==' is a comparison operator, which we will discuss later.

### For this exercise please do the following in the cell below.
1. Run the comparison code above and print the result. If you don't believe the answer, change '0.j' to '1j' and run the code again.
1. Type-cast 0.5 and 0.99 to integers and print the results
1. Similar to above, but now type-cast 100.5 and 100.99 to integers
1. Set $x$ equal to a floating point number with a non-zero value left of the decimal point. Then finish the line below such that $x$ is rounded to the nearest integer no mater what value is to the right of the decimal point. Print that result.
1. Store the results of the following statements into a list (*hint: use .append()*). After that, run the code for section 5 below which maps each element in the list to it's corresponding boolean value. Using comments (i.e. #comment), discuss the results. 
```python 
True
True + 1
True - 1
True*1
True*0
True*3.1425
True*-3.1425
False
False + 1
False - 1
False*1
False*0
False*3.1425
False*-3.1425
```

In [2]:
# 1. comparison
print('#1')
is_same = (z*y == z*(y + 0.j))
print(f'Equivalent?: {is_same}')
w = y + 1j
print(w)
print(f'Equivalent?: {z*w == z*(y + 1.j)}')

# 2. type-cast
print('\n#2')
ex1p2_typecast = lambda num : print(f'Type cast of int({str(num)}) = {int(num)}\n')
ex1p2_typecast(0.5)
ex1p2_typecast(0.99)

# 3. type-cast
print('\n#3')
ex1p2_typecast(100.5)
ex1p2_typecast(100.99)

# 4. round to nearest integer (you need to add something before casting)
print('\n#4')
ex1p4_rounded = lambda num : int(num + 0.5)
ex1p4_print_rounded = lambda num : print(f'Manual Round({str(num)}) = {ex1p4_rounded(num)}\n')

for num in [0.5, 0.99, 100.5, 100.99, 104.499, 104.01]:
    ex1p4_print_rounded(num)


# 5. 'Boolean' is a number
print('\n#5')
## name your list: mylist
mylist = [
    True,
    True + 1,
    True - 1,
    True*1,
    True*0,
    True*3.1425,
    True*-3.1425,
    False,
    False + 1,
    False - 1,
    False*1,
    False*0,
    False*3.1425,
    False*-3.1425]

# your code

print(f'mylist: {mylist}')

myboollist = list(map(bool,mylist))
print(f'bool(mylist): {myboollist}')

# Table-wise printing
mylist_str = [
    "True",
    "True + 1",
    "True - 1",
    "True*1",
    "True*0",
    "True*3.1425",
    "True*-3.1425",
    "False",
    "False + 1",
    "False - 1",
    "False*1",
    "False*0",
    "False*3.1425",
    "False*-3.1425"]

myziplist = zip(mylist_str, mylist, myboollist)
for triple in myziplist:
    print(triple)
    
# Discussing the results:
# One might think that 0 <-> False and 1 <-> True is biconditional, as I did. 
# Unfortunately, this is not the case for True. In particular, for False, we
# have that False only maps to 0 and 0 only maps to False. But, for True, we
# have that True only maps to 1, BUT any non-zero number (integer or float)
# will map to True. 

#1
Equivalent?: True
(1.2+1j)
Equivalent?: True

#2
Type cast of int(0.5) = 0

Type cast of int(0.99) = 0


#3
Type cast of int(100.5) = 100

Type cast of int(100.99) = 100


#4
Manual Round(0.5) = 1

Manual Round(0.99) = 1

Manual Round(100.5) = 101

Manual Round(100.99) = 101

Manual Round(104.499) = 104

Manual Round(104.01) = 104


#5
mylist: [True, 2, 0, 1, 0, 3.1425, -3.1425, False, 1, -1, 0, 0, 0.0, -0.0]
bool(mylist): [True, True, False, True, False, True, True, False, True, True, False, False, False, False]
('True', True, True)
('True + 1', 2, True)
('True - 1', 0, False)
('True*1', 1, True)
('True*0', 0, False)
('True*3.1425', 3.1425, True)
('True*-3.1425', -3.1425, True)
('False', False, False)
('False + 1', 1, True)
('False - 1', -1, True)
('False*1', 0, False)
('False*0', 0, False)
('False*3.1425', 0.0, False)
('False*-3.1425', -0.0, False)


## Exercise 2: "Type-casting" Numpy arrays

Type casting Numpy arrays is a little different than it is for built in types.  Here, we can make use of the following function, [**astype()**](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.astype.html).

### For this exercise please do the following in the cell below.
1. Run the code below for this problem below to create an array of random 64bit floats. Then type cast the array to float32 types, and then subtract the two arrays and print the results. Discuss the result of the subtraction.
1. Now, using the same 64bit array, type cast all the values to int32 values. How were the numbers rounded? Was this what you expected compared to rouding a Python built in type?
1. Now type cast the float64 array to the nearst int32.
1. Dynamically cast the int32 array created above to into an array with float64 types. Print the array and something that verifies the arrays new numeric type.
1. For the last problem in this section. Type cast the float64 array created in problem 1, to an array of unsigned integers ([Hint](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.uint32)). What do you notice with the results compared to problem 2?

In [3]:
import numpy as np

# 1 type cast to float32
print('#1')
np.random.seed(42) # leave seed alone until all 5 problems and discussion are finished, then feel free to tinker.
isc = np.random.randint(5,high=10)
rx64 = (np.random.rand(10)-0.5)*isc
# your code
print(f'Base 64bit floats = {rx64}')
rx32 = rx64.astype(np.float32)
print(f'Casting to 32bit floats = {rx32}')

sub_rx = rx64 - rx32
print(f'Subtracted array = {sub_rx}')
print(f'Subtracted array is of type {sub_rx.dtype}')

# As expected, numpy casted the float32 to a float64 during the subtraction.
# It roughly, but not exactly, rounded off the decimals at the seventh 
# decimal place. Note that the differences are both positive and negative,
# illustrating that this is also not a floor or ceiling operation. 

# 2 type cast to int32
print('\n#2')
# your code
ix32 = rx64.astype(np.int32)
print(f'Base 64bit floats = {rx64}')
print(f'Casting to 32bit ints = {ix32}')
# Similar to the built-in Python versions, numpy rounded to integer by flooring
# the decimal place (i.e., removing the decimal part and only keeping the 
# integer part). 


# 3 type cast to nearest int32
print('\n#3')
# your code
ix32_round = (rx64 + 0.5 * np.sign(rx64)).astype(np.int32)
print(f'Base 64bit floats = {rx64}')
print(f'Rounding to 32bit ints = {ix32_round}')

# 4 dynamic type cast int32 array above to float64
print('\n#4')
# your code
rx64_dynamic = 1.0 * ix32
print(f'Base 64bit floats = {rx64}')
print(f'Casted 32bit ints = {ix32}')
print(f'Dynamic Cast to 64bit floats = {rx64_dynamic}')
print(f'Type after dynamic = {rx64_dynamic.dtype}')

# 5 type cast to unsigned-int32
print('\n#5')
# your code
uix64 = rx64.astype(np.uint64)
uix64_negative = (-rx64).astype(np.uint64)
print(f'Base 64bit floats = {rx64}')
print(f'Casting to 32bit unsigned ints = {uix64}')
print(f'Casting Negative to 32bit unsigned ints = {uix64_negative}')
# Because the unsigned integers can only represent positive integers, 
# numpy wraps from 0 to the largest integer and subtracts from there.
# The largest unsigned integer is 2^64 - 1 = (18,446,744,073,709,551,615)
# The behavior is especially clear when we negate the array and cast.

#1
Base 64bit floats = [ 3.60571445  1.85595153  0.78926787 -2.75185088 -2.75204384 -3.5353311
  2.92940917  0.80892009  1.66458062 -3.83532405]
Casting to 32bit floats = [ 3.6057146  1.8559515  0.7892679 -2.7518508 -2.7520437 -3.535331
  2.9294093  0.8089201  1.6645806 -3.835324 ]
Subtracted array = [-1.08275724e-07 -1.31314399e-08 -2.40296032e-08 -3.30309424e-08
 -1.13250320e-07 -9.18359229e-08 -9.93187070e-08  8.51552517e-09
  3.87959762e-08  3.36239125e-09]
Subtracted array is of type float64

#2
Base 64bit floats = [ 3.60571445  1.85595153  0.78926787 -2.75185088 -2.75204384 -3.5353311
  2.92940917  0.80892009  1.66458062 -3.83532405]
Casting to 32bit ints = [ 3  1  0 -2 -2 -3  2  0  1 -3]

#3
Base 64bit floats = [ 3.60571445  1.85595153  0.78926787 -2.75185088 -2.75204384 -3.5353311
  2.92940917  0.80892009  1.66458062 -3.83532405]
Rounding to 32bit ints = [ 4  2  1 -3 -3 -4  3  1  2 -4]

#4
Base 64bit floats = [ 3.60571445  1.85595153  0.78926787 -2.75185088 -2.75204384 -3.53533

## Python and Numpy Overflow Errors (integers)

The behavior you just saw in problem 5 above is related to a type of computational error called an Overflow, specifically, a Binary Overflow (not related to a "Stack Overflow" which is a real thing, and not just a website for finding solutions to coding problems). I think the following is a reasonable discription of a [Binary Overflow](https://www.geeksforgeeks.org/overflow-in-arithmetic-addition-in-binary-number-system/).

### Exercise 3

For this exercise, extend the code below to int64 type integers (copy and modify including the comments). Then in the markdown cell for this exercise below, please explain:
1. Why does adding *1* to the numpy varibales cause the values to become negative, while the same operation does not cause the Python variable values to become negative. ([Hint-1](https://docs.python.org/3.3/reference/lexical_analysis.html#numeric-literals), [Hint-2](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.int32))
1. Suppose you are writting a specific code, and you know that none of the operations will result in negative values, what other interger type (dtype) could use for 32-bit and 64-bit numpy arrays that would extend the maximum interger value that the arrays could correctly represent? ([Hint](https://programmercave0.github.io/blog/2019/10/19/Bit-Manipulation-in-C-and-C++))
1. Why is a Numpy array '+=' operator being used in Problem 1 instead of just doing something like the code shown below? Try this code your self, and then modify if so that you can understand and explain what happens after the addition of $1$ to the Numpy variable vs. the addition to the Numpy array.
```python
np_x32 = np.array([2**31-1],dtype=np.int32)[0]
#                               look here --^
```
```python
np_x32 += 1 #Single variable += operator
print(f'val(np_x32+1), type(np_x32+1): {np_x32},{type(np_x32)}')
#                         what is not here --^  nor here --^
```


In [4]:
# Python int32 
print('Python int32')
x32 = int(2**31 - 1) #remember order of operations!
print(f'val(x32), x32.bit_length: {x32},{x32.bit_length()}')

x32 += 1
print(f'val(x32+1), (x32+1).bit_length: {x32},{x32.bit_length()}')

print()

# Numpy int32 
print('Numpy int32')
np_x32 = np.array([2**31-1],dtype=np.int32)
print(f'val(np_x32), type(np_x32): {np_x32[0]},{type(np_x32[0])}')

np_x32 += 1
print(f'val(np_x32+1), type(np_x32+1): {np_x32[0]},{type(np_x32[0])}')

print()

# Python int64 
print('Python int64')

#   Your Code Here
x64 = int(2**63 - 1) #remember order of operations!
print(f'val(x64), x64.bit_length: {x64},{x64.bit_length()}')

x64 += 1
print(f'val(x64+1), (x64+1).bit_length: {x64},{x64.bit_length()}')

print()

# Numpy int64 
print('Numpy int64')

#   Your Code Here
np_x64 = np.array([2**63-1],dtype=np.int64)
print(f'val(np_x64), type(np_x64): {np_x64[0]},{type(np_x64[0])}')

np_x64 += 1
print(f'val(np_x64+1), type(np_x64+1): {np_x64[0]},{type(np_x64[0])}')

print()

# See discussion below.

# Problem 2: 
print('Numpy uint32')
np_x_uint_32 = np.array([2**31-1],dtype=np.uint32)
print(f'val(np_x_uint_32), type(np_x_uint_32): {np_x_uint_32[0]},{type(np_x_uint_32[0])}')

np_x_uint_32 += 1
print(f'val(np_x_uint_32+1), type(np_x_uint_32+1): {np_x_uint_32[0]},{type(np_x_uint_32[0])}')

print()

np_x_uint_32_max = np.array([2**32-1],dtype=np.uint32)
print(f'val(np_x_uint_32_max), type(np_x_uint_32_max): {np_x_uint_32_max[0]},{type(np_x_uint_32_max[0])}')

np_x_uint_32_max += 1
print(f'val(np_x_uint_32_max+1), type(np_x_uint_32_max+1): {np_x_uint_32_max[0]},{type(np_x_uint_32_max[0])}')

print()


#Problem 3: Varible Numpy int32

#   Your Code Here
print('\n\n#3')
np_x32 = np.array([2**31-1],dtype=np.int32)[0]
#                               look here --^
print(f'val(np_x32), type(np_x32): {np_x32},{type(np_x32)}')
np_x32 += 1 #Single variable += operator
print(f'val(np_x32+1), type(np_x32+1): {np_x32},{type(np_x32)}')
#                         what is not here --^  nor here --^

Python int32
val(x32), x32.bit_length: 2147483647,31
val(x32+1), (x32+1).bit_length: 2147483648,32

Numpy int32
val(np_x32), type(np_x32): 2147483647,<class 'numpy.int32'>
val(np_x32+1), type(np_x32+1): -2147483648,<class 'numpy.int32'>

Python int64
val(x64), x64.bit_length: 9223372036854775807,63
val(x64+1), (x64+1).bit_length: 9223372036854775808,64

Numpy int64
val(np_x64), type(np_x64): 9223372036854775807,<class 'numpy.int64'>
val(np_x64+1), type(np_x64+1): -9223372036854775808,<class 'numpy.int64'>

Numpy uint32
val(np_x_uint_32), type(np_x_uint_32): 2147483647,<class 'numpy.uint32'>
val(np_x_uint_32+1), type(np_x_uint_32+1): 2147483648,<class 'numpy.uint32'>

val(np_x_uint_32_max), type(np_x_uint_32_max): 4294967295,<class 'numpy.uint32'>
val(np_x_uint_32_max+1), type(np_x_uint_32_max+1): 0,<class 'numpy.uint32'>



#3
val(np_x32), type(np_x32): 2147483647,<class 'numpy.int32'>
val(np_x32+1), type(np_x32+1): 2147483648,<class 'numpy.int64'>


<br><br>
### Response to Exercise 3 in this *Markdown* cell
<br><br>

[//]: <> (Delete one of the <br> tags from above the heading but leave the rest.)
[//]: <> (Then fill in your responses and number your responses via a numbered, markdown list.)
[//]: <> (Also, please give me your oppinion on who would win in a battle between Superman and Batman.)
[//]: <> (Make the heading of your Opinion just one-size smaller then the heading of this cell.)
[//]: <> (One sentence or even just one word about the battle will suffice)

1. Python Integers are variable length. In this case, it lengthened from an integer specified with a 31 bit length to an integer with a 32 bit length. In contrast, NumPy Integers have a fixed size. So, for example, we specified a 32 bit integer (where one bit is used for the sign). So, when we requested to add, it overflowed by one bit and looped around to the negative numbers. 
2. We can use unsigned integers. Using an unsigned integer doubled the size of the maximum integer value. In doing so, it overflows to zero rather than to negative numbers.
3. Python and NumPy are smart, but subtle. When we store the variable asan array, NumPy has to maintain the same type for each variable and store each element with the allocated memory footprint of the given type. NumPy is trying to avoid the case where one value of a large array is about to overflow and then decide that it needs to upgrade the whole array from int32 to int64. That would generally be a big waste of memory. However, if we only have one value, NumPy is willing to make the change and request more memory. So, it dynamically casts the number from int32 to int64. 

#### Superman vs Batman
I think Superman would win since Batman is unlikely to acquire Kryptonite befor Superman destroys Batman.

## Numpy Underflow Errors (floating point)

Now that you have seen Overflow errors, lets look at what Underflow errors are.  Do you have a guess?  The following [Wiki](https://en.wikipedia.org/wiki/Arithmetic_underflow) provides a nice explination of this type of error. There is more information on that page then you probably need to know, but be sure to read the first section. Then work on the exercise below.

### Exercise 4

For this exercise, besure to comment the code, and provide print statments between the problems.  Make the output look nice. For the questions, please fill in your responses in the markdown section that follows.

* Part A
    0. Print the smallest possible numpy.float32 value. (This has been done for you.) [Extra-Info](https://en.wikipedia.org/wiki/Subnormal_number)
    1. print the result of mf$32^2$. Do you believe that the value in A0 really is the smallest numpy.float32? Provide evidence that supports your answer (Hint: this is possible with one line of code.)
    2. Now creat a numpy array, np_mf32, with a size one and the value of mf32 in it. Print both the array and the numpy.dtype of the array. (no discussion needed)
    3. Now square the the np_mf32 array.  Does this confirm your thoughts about A1?
    4. Create a list or numpy array of floats for $x$ such that $ x \in \left[0.5,0.9\right]$ by increments of 0.1 (there shold be 5 elements). No loop over the values of $x$ and multiply the np_mf32 array by x. Comment on which value or values cause an Underflow error.
* Part B: do all of the above in Part A with a numpy.float64 dtype
    5. And, answer this question related to A1 vs B1. Why were the results differen when we squared mf32 in A1 vs squaring mf64 in B1? (Hint: what is the dtype of both values after squaring? It takes just two lines of code to see what was different between the two.)


In [5]:
#Part A

#A0
print("#A0 (Part A, Problem 0)")
#This should be the smallest float32 possible, but ...
mf32 = np.finfo(np.float32).smallest_subnormal 
print(f'val(mf32): {mf32}, type = {type(mf32)}')

#A1 square mf32
print("\n#A1 (Part A, Problem 1)")
mf32_squared = mf32 ** 2
print(f'val(mf32_squared): {mf32_squared}, type = {type(mf32_squared)}')

def a0smaller32(delta, val=mf32):
    '''Attempt to go below "val" and keep it a float32.'''
    # Check the input value is int32
    assert(isinstance(mf32, np.float32))
    
    # Cast to float32, otherwise the difference will recast it
    delta_float32 = np.float32(delta)
    assert(isinstance(delta_float32, np.float32))
    
    # Now subtract
    new_val = val - delta_float32
    print(f'delta of {delta} converted to float32 as {delta_float32}')
    print(f'val(mf32 - delta): {new_val}, type = {type(new_val)}')
    return

print("\nAttempting to go smaller than what NumPy says")
a0smaller32(1e-44)
a0smaller32(3e-45)
a0smaller32(1e-45)

#A2 mf32 in a numpy array (look at the Numpy int32 example in the previous exercise)
print("\n#A2 (Part A, Problem 2)")
np_mf32 = np.array([mf32], dtype=np.float32)
print(f'mf32 as array = {np_mf32} of type {type(np_mf32)}.')
print(f'mf32 element in array = {np_mf32[0]} of type {type(np_mf32[0])}.')

#A3 square the np_mf32 array 
print("\n#A3 (Part A, Problem 3)")
np_mf32_squared = np_mf32 ** 2
print(f'np_mf32 ** 2 = {np_mf32_squared} of type {type(np_mf32_squared)}.')
print(f'zeroth element of array = {np_mf32_squared[0]} of type {type(np_mf32_squared[0])}.')

#A4 loop over x in [0.5:0.9:0.1] show x*np_mf32
print("\n#A4 (Part A, Problem 4)")
a4smaller = lambda ratio, val : print(f'val(mf32 * {ratio}) = {(ratio * val)[0]}, type = {type((ratio * val)[0])})')
x = np.arange(0.5, 1.0, 0.1)
print(x)
for xi in x:
    a4smaller(xi, np_mf32)
    
x = np.arange(0.498, 0.502, 0.001)
print(x)
for xi in x:
    a4smaller(xi, np_mf32)

#A0 (Part A, Problem 0)
val(mf32): 1.401298464324817e-45, type = <class 'numpy.float32'>

#A1 (Part A, Problem 1)
val(mf32_squared): 1.9636373861190906e-90, type = <class 'numpy.float64'>

Attempting to go smaller than what NumPy says
delta of 1e-44 converted to float32 as 9.80908925027372e-45
val(mf32 - delta): -8.407790785948902e-45, type = <class 'numpy.float32'>
delta of 3e-45 converted to float32 as 2.802596928649634e-45
val(mf32 - delta): -1.401298464324817e-45, type = <class 'numpy.float32'>
delta of 1e-45 converted to float32 as 1.401298464324817e-45
val(mf32 - delta): 0.0, type = <class 'numpy.float32'>

#A2 (Part A, Problem 2)
mf32 as array = [1.e-45] of type <class 'numpy.ndarray'>.
mf32 element in array = 1.401298464324817e-45 of type <class 'numpy.float32'>.

#A3 (Part A, Problem 3)
np_mf32 ** 2 = [0.] of type <class 'numpy.ndarray'>.
zeroth element of array = 0.0 of type <class 'numpy.float32'>.

#A4 (Part A, Problem 4)
[0.5 0.6 0.7 0.8 0.9]
val(mf32 * 0.5) = 0.0, type = 

In [6]:
# Part B (same as above but for float64)


#B0 - B4: Your code

# First, we will demonstrate what happens to mf32 as we go smaller and smaller
# Illustrate that with float64 we can go WAY smaller
a0smaller = lambda ratio, val : print(f'val(mf32 * {ratio}) = {ratio * val}, type = {type(ratio * val)})')

print("\nAttempting to underflow the float64 starting at mf32")
a0smaller(1e-46, mf32) # We can make a number smaller than the square and keep it a float64
print()
a0smaller(1e-211, mf32) # We would think we overflow the 8bit exponent width for float64 (max value of 255)
a0smaller(1e-264, mf32) # We do not see any issue in representaton even when we multiply by 1e-264
a0smaller(1e-265, mf32) # Finally we see issues at 1e-265 in the smallest two decimal points
a0smaller(1e-270, mf32) # The issues are much more noticeably by 1e-270...
a0smaller(1e-275, mf32) # and much more dramatic by 1e-275
a0smaller(1e-276, mf32) # Now we start losing roughly one decimal point per 0.1x reduction
a0smaller(1e-277, mf32)
a0smaller(1e-278, mf32) # This is the smallest exponent we cna multiply with unit coefficient and integer exponent
a0smaller(1e-279, mf32) # We finally underflow.

# B0
print("\n#B0")
mf64 = np.finfo(np.float64).smallest_subnormal
print(f'val(mf64): {mf64}, type = {type(mf64)}')

# B1
print("\n#B1")
mf64_squared = mf64 ** 2
print(f'val(mf64 ** 2) = {mf64_squared}, type = {type(mf64_squared)}')

#B2 mf64 in a numpy array
print("\n#B2 (Part B, Problem 2)")
np_mf64 = np.array([mf64], dtype=np.float64)
print(f'mf64 as array = {np_mf64} of type {type(np_mf64)}.')
print(f'mf64 element in array = {np_mf64[0]} of type {type(np_mf64[0])}.')

#B3 square the np_mf64 array 
print("\n#B3 (Part B, Problem 3)")
np_mf64_squared = np_mf64 ** 2
print(f'np_mf64 ** 2 = {np_mf64_squared} of type {type(np_mf64_squared)}.')
print(f'zeroth element of array = {np_mf64_squared[0]} of type {type(np_mf64_squared[0])}.')

#B4 loop over x in [0.5:0.9:0.1] show x*np_mf64
print("\n#B4 (Part A, Problem 4)")
b4smaller = lambda ratio, val : print(f'val(mf64 * {ratio}) = {(ratio * val)[0]}, type = {type((ratio * val)[0])})')
x = np.arange(0.5, 0.9, 0.1)
print(x)
for xi in x:
    b4smaller(xi, np_mf64)
    
x = np.arange(0.5, 0.505, 0.001)
print(x)
for xi in x:
    b4smaller(xi, np_mf64)

#B5 The extra qustion: A1 vs B1
# Already completed above.


Attempting to underflow the float64 starting at mf32
val(mf32 * 1e-46) = 1.401298464324817e-91, type = <class 'numpy.float64'>)

val(mf32 * 1e-211) = 1.4012984643248172e-256, type = <class 'numpy.float64'>)
val(mf32 * 1e-264) = 1.401298464324817e-309, type = <class 'numpy.float64'>)
val(mf32 * 1e-265) = 1.40129846432483e-310, type = <class 'numpy.float64'>)
val(mf32 * 1e-270) = 1.401298466e-315, type = <class 'numpy.float64'>)
val(mf32 * 1e-275) = 1.401e-320, type = <class 'numpy.float64'>)
val(mf32 * 1e-276) = 1.403e-321, type = <class 'numpy.float64'>)
val(mf32 * 1e-277) = 1.4e-322, type = <class 'numpy.float64'>)
val(mf32 * 1e-278) = 1.5e-323, type = <class 'numpy.float64'>)
val(mf32 * 1e-279) = 0.0, type = <class 'numpy.float64'>)

#B0
val(mf64): 5e-324, type = <class 'numpy.float64'>

#B1
val(mf64 ** 2) = 0.0, type = <class 'numpy.float64'>

#B2 (Part B, Problem 2)
mf64 as array = [5.e-324] of type <class 'numpy.ndarray'>.
mf64 element in array = 5e-324 of type <class 'numpy.floa

<br><br><br>
### Response to Exercise 4, parts A and B in this *Markdown* cell
<br><br>

[//]: <> (Make this look decent. Use #### sized headings for the Parts, and a numbered markdown list for the responses.)
[//]: <> (Also, I think Batman would win, as long as he isn't taken by surprise.)

#### Part A
0. Apparently, the smallest `numpy.float32` is `1.401298464324817e-45`, according to numpy.
1. As far as I can see, the value in A0 (above) is the smallest `numpy.float32` that we can represent, albeit way larger than the smallest `numpy.float64` that we can represent (see Part B). 
2. None
3. Now, when we square the number, we underflow immediately and do not typecast to float64. Like the previous problem, it seems that numpy does not want to use the typecast approach for arrays. I hypothesize that the design choice is similar to above. If we typecast `float32` to `float64` to have values smaller than `smallest_subnormal` as `float32`, we will have to change the allocation and double the memory footprint, which could be expensive and annoying for large arrays.
4. The underflow occurs at 0.5 (exactly) since this is 1/2 of the smallest number and we have passed the gap between floats.


#### Part B
To make sense of what is happening with `numpy.float64`, I started by reducing the exponent from mf32. We can actually multiply mf32 (AKA the value in A0) by $10^{-278}$ before we cannot represent it anymore. However, we begin losing precision in the decimal point. To illustrate this, I included $10^{-275}$ through $10^{-279}$. This matches the description on the linked wikipedia page:
> By filling the underflow gap like this, significant digits are lost, but not as abruptly as when using the flush to zero on underflow approach (discarding all significant digits when underflow is reached). Hence the production of a subnormal number is sometimes called gradual underflow because it allows a calculation to lose precision slowly when the result is small.
> https://en.wikipedia.org/wiki/Subnormal_number

With that, the direct responses are below.

0. The smallest `numpy.float64` is `5e-324`. 
1. In part A, numpy promoted the float from `numpy.float32` to `numpy.float64`. However, numpy does not do this with `numpy.float64`. Namely, it does not promote from `numpy.float64` to `numpy.float128`. Instead, it immediately underflows.
2. None
3. In this case, the array has the same behavior as the number alone. The array also immediately overflows without typecasting. 
4. Again, the underflow occurs at 0.5 since this is 1/2 of the smallest number and we have passed the gap between floats.
5. Ultimately, this is a design decision. Numpy is willing to promote from `numpy.float32` to `numpy.float64`, but it is not willing to promote from `numpy.float64` to `numpy.float128`. Maybe if we have larger memory or have a breakthrough with memory management, numpy will start allowing operations to promote by default to `numpy.float128`.

#### Superman vs. Batman
I do agree that Batman is perhaps more clever and cunning than Superman. So, with enough preparation, I think Batman beats Superman in the long game. 

## Round-off/Truncation Errors (floating point)

Before starting this exercise, please read the [Round-off Errors](https://pythonnumericalmethods.berkeley.edu/notebooks/chapter09.03-Roundoff-Errors.html) section in Chapter 9 of *Python Numerical Methods*.

### Exercise 5

For this exercise, were are going to examine [Truncation](http://nifty.stanford.edu/2003/pests/2002/lectures/07.1_FloatingPoint/truncation.htm?CurrentSlide=6) errors, which are related to Round-off errors. An example of when you might encounter truncation errors is when solving integrals (a sequence of summations) on a computer. Given that we mostly work with computers that have finite precision arithematic, we must represent integrals as finite summations.

To demonstrate truncation errors we are going to sum all the integer values from $1$ to $N$, but were are going to represent each integer as a numpy.float32 type. In the cell below, I've writen a function that finds the analyical solution to $\sum\limits_{i=1}^{N}i$, for a given $N$.

Your job (write responses in the markdown cell below):
1. First, create a Numpy array of size $N$ containing numpy.float32 values representing the integers from $1-N$, and have the values stored in consective indexing order (i.e. arr[0] = 1.0, arr[1] = 2.0, ..., arr[N-1] = N.0).
2. Next, you need to create function that calculates and returns the sum of this array (make it fast, and feel free to use numpy functions). Besure to verify that your sum is the same as the analytical sum. I suggest starting with an array of just 10 elements, so $N=10$.
3. Once you have verified that your sumation function is returning the correct results, you have two tasks:
    1. Find the value of $N_{inc}$ at which your summation is no longer corret. Discuss in the markdown cell below how you found this value. 
    2. Write an improved summation that can correctly sum all the values to find the answer of $\sum\limits_{i=1}^{2\cdot N_{inc}}$ (the sum upto $2\cdot N_{inc}$). Discuss your reasoning about how you chose to solve this task.
    3. Special Case: If you are taking this course for 4-units, then write a recursive function recur_sum() that can solve problem 2. All values must remain as numpy.float32 values (no type-casting). Then test your fuction for $4\cdot N_{inc}$ (it should be accurate without any changes, but if it's not, that's ok).



In [7]:
import numpy as np


# a function that returns the analytical sum of the numbers 1 through N
def analytic_sum(N, ftype = np.float32):
    # I added 'ftype' as an arguement since 'ftype' seems to be undefined otherwised
    return np.array([0.5*N*(N+1)]).astype(np.dtype(ftype))[0]

# PART 1
def gen_np_n_arr(N):
    """Generate an array of numbers from 1.0 to N as numpy.float32"""
    return np.arange(1.0, N + 1, dtype=np.float32)

# PART 2
# your summation function that takes numpy array of type numpy.float32
def my_sum(myarr):
    #your code
    # We initialize the sum total to zero
    total = np.float32(0.0)
    for i in myarr:
        total += i
    return total

# PART 3
def find_sum_breakpoint(func_sum, Ninitial=1000, tol=0.0, max_iters=100):
    """
    Find the first number where the summation fails using binary search with logarithmic initialization.
    
    Inputs:
    - func_sum   = The function that we are using to perform the summation 
    - Ninitial   = The initial number to begin the search
    - tol        = The tolerance where we declare the sum failed
    """
    # Here we set a loose lower bound as N_works and the upper bound as N_failure
    N_works = 10
    N_failure = np.inf
    
    N_curr = Ninitial
    
    
    for iter_num in range(max_iters) :
        
        print(f"ITERATION {iter_num}: Current N is {N_curr}")
        # Generate the sequence
        seq = gen_np_n_arr(N_curr)
        # Compute the sum
        test_sum = func_sum(seq)
        # Compute the analytical sum
        analytic_sum_val = analytic_sum(N_curr)
        
        # Compute the difference between the computed series and analytical solution
        diff_abs = np.abs(test_sum - analytic_sum_val)
        
        # Show the computation
        print(f'Test Sum = {test_sum}. Analytical Sum = {analytic_sum_val}. |Diff| = {diff_abs}')
        
        # Update the bounds
        if diff_abs > tol:
            # The sum failed
            N_failure = N_curr
            # Find the middle of N_failure and N_works
            N_curr = int((N_failure + N_works) / 2)
        else:
            # The sum passed
            N_works = N_curr
            
            if np.isinf(N_failure):
                N_curr *= 10
            else:
                N_curr = int((N_failure + N_works) / 2)
        
        err_msg = f'N_failure = {N_failure} and N_curr = {N_curr} and N_works {N_works}'
        assert N_failure >= N_curr, err_msg
        assert N_curr >= N_works, err_msg
        
        # While N_failure is much greater than N_works, there must be tighter bounds
        if (N_failure - N_works) < 2:
            break
    
    # Return the breakpoint
    return N_works
    

# your improved summation function 
def my_impr_sum(myarr, split_num=32):
    #your code
    # We split the summation into smaller summations and add the totals
    if split_num > len(myarr):
        split_num = len(myarr)
    splitted_arrs = np.array_split(myarr, split_num)
    
    total = np.float32(0.0)
    for arr_split in splitted_arrs:
        total += my_sum(arr_split)
    
    return total

# Special Case: recursive sum function here
def my_recursive_sum(myarr):
    """We approach summation as a divide and conquer algorithm."""
    if len(myarr) == 1:
        # Base Case
        return myarr[0]
    
    # Split the array in two and add the splits
    splitted_arrs = np.array_split(myarr, 2)
    tot_left = my_recursive_sum(splitted_arrs[0])
    tot_right = my_recursive_sum(splitted_arrs[1])
    return tot_left + tot_right
        

# put your code to test the functions here. Use comments!

# TESTING PART 1
print('\n#1')
# Generate the numpy array for N=10
n1 = 10
arr_np_to_n = gen_np_n_arr(n1)
print(f'Array is {arr_np_to_n} of type {arr_np_to_n.dtype}')

# TESTING PART 2
# Sum up the numbers 
print('\n#2')
tot1 = my_sum(arr_np_to_n)
print(f'Array sum is {tot1}. The analytical sum is {analytic_sum(n1)}')

# PART 3A
print('\n#3A')
breakpt3A = find_sum_breakpoint(my_sum)
print(f'\n\nThe breakpoint for accurately computing the sum (via my_sum) is {breakpt3A}\n\n')

breakpt3A_npsum = find_sum_breakpoint(np.sum)
print(f'\n\nThe breakpoint for accurately computing the sum (via np.sum) is {breakpt3A_npsum}\n\n')

# PART 3B
breakpt3B = find_sum_breakpoint(my_impr_sum)
print(f'\n\nThe breakpoint for accurately computing the sum (via my_impr_sum) is {breakpt3B}\n\n')
print(f'Comparison: my_impr_sum is {breakpt3B / breakpt3A} times better than my_sum,')
print(f' but {breakpt3A_npsum / breakpt3B} times worse than np.sum.\n\n')

# PART 3C
# Check that it works
# print(my_recursive_sum(arr_np_to_n))

breakpt3C = find_sum_breakpoint(my_recursive_sum)
print(f'\n\nThe breakpoint for accurately computing the sum (via my_recursive_sum) is {breakpt3C}\n\n')
print(f'Comparison: my_recursive_sum is {breakpt3C / breakpt3A} times better than my_sum,')
print(f' but {breakpt3A_npsum / breakpt3C} times worse than np.sum.')

# Print the final results
print('\n\n')
print(f'     FUNCTION      |      METHOD      |   MAX N  ')
print(f'-------------------------------------------------')
print(f'      my_sum       |    Direct Sum    | {breakpt3A}')
print(f'    my_impr_sum    |   32-split Sum   | {breakpt3B}')
print(f' my_recursive_sum  | 2-way Div & Conq | {breakpt3C}')
print(f'      np.sum       | Partial Pairwise | {breakpt3A_npsum}')


#1
Array is [ 1.  2.  3.  4.  5.  6.  7.  8.  9. 10.] of type float32

#2
Array sum is 55.0. The analytical sum is 55.0

#3A
ITERATION 0: Current N is 1000
Test Sum = 500500.0. Analytical Sum = 500500.0. |Diff| = 0.0
ITERATION 1: Current N is 10000
Test Sum = 50002896.0. Analytical Sum = 50005000.0. |Diff| = 2104.0
ITERATION 2: Current N is 5500
Test Sum = 15127750.0. Analytical Sum = 15127750.0. |Diff| = 0.0
ITERATION 3: Current N is 7750
Test Sum = 30034146.0. Analytical Sum = 30035124.0. |Diff| = 978.0
ITERATION 4: Current N is 6625
Test Sum = 21948208.0. Analytical Sum = 21948624.0. |Diff| = 416.0
ITERATION 5: Current N is 6062
Test Sum = 18376818.0. Analytical Sum = 18376952.0. |Diff| = 134.0
ITERATION 6: Current N is 5781
Test Sum = 16712871.0. Analytical Sum = 16712871.0. |Diff| = 0.0
ITERATION 7: Current N is 5921
Test Sum = 17532016.0. Analytical Sum = 17532080.0. |Diff| = 64.0
ITERATION 8: Current N is 5851
Test Sum = 17119996.0. Analytical Sum = 17120026.0. |Diff| = 30.0
IT

Test Sum = 2454166784.0. Analytical Sum = 2454166784.0. |Diff| = 0.0
ITERATION 19: Current N is 70060
Test Sum = 2454236928.0. Analytical Sum = 2454236928.0. |Diff| = 0.0


The breakpoint for accurately computing the sum (via my_recursive_sum) is 70060


Comparison: my_recursive_sum is 12.093906438805455 times better than my_sum,
 but 5.843805309734513 times worse than np.sum.



     FUNCTION      |      METHOD      |   MAX N  
-------------------------------------------------
      my_sum       |    Direct Sum    | 5793
    my_impr_sum    |   32-split Sum   | 12912
 my_recursive_sum  | 2-way Div & Conq | 70060
      np.sum       | Partial Pairwise | 409417


<br><br>
### Responses to Exercise 5
<br><br>

[//]: <> (Make this look decent. Use a numbered markdown list for the responses to each task.)

Responses for Part 3
1. The direct sum (`my_sum`) is no longer correct after 5793 values to sum up.
2. Looking at the `np.sum` documentation and given that the next problem recommends a recursive solution, the idea is to split the giant series into batches of smaller series. In `my_impr_sum`, I split the sum into 32 smaller arrays. This way, we contain the round-off error within the 32 smaller series rather than propagating the sum through the entire summation. With this strategy, we can reach a maximum $N_{inc}$ of 12912, which is 2.2 times better than the original direct summation strategy.
3. Even though I am taking the class for 1 unit, this problem seemed interesting, so I took a try at it. Here I use a full divide and conquer algorithm. I recursively split the array into 2 smaller arrays. Once we have a single element, we return the element. We then compute the sums across base cases as a tree. So, round-off can only impact a single branch of the tree. The implementation is slow and naive, but the aim is beating 4 times, which we achieve. Specifically, we can reach a maximum $N_{inc}$ of 70060, which is 12.1 times better than the original direct summation strategy. However, we are still 5.84 times worse than `np.sum`, so there must be other algorithmic tricks in `np.sum`.